SimFastTiming/FastTimingCommon/src/BTLElectronicsSim.cc

```diff
@@ -2,44 +2,123 @@
+  SigmaClock2_(SigmaClock_*SigmaClock_) { 
 }
 
 
 void BTLElectronicsSim::run(const mtd::MTDSimHitDataAccumulator& input,
-			    BTLDigiCollection& output) const {
-  
-  MTDSimHitData chargeColl,toa;
+			    BTLDigiCollection& output,
+			    CLHEP::HepRandomEngine *hre) const {
   
+  MTDSimHitData chargeColl, toa1, toa2;
+
   for(MTDSimHitDataAccumulator::const_iterator it=input.begin();
       it!=input.end();
       it++) {
     
-    chargeColl.fill(0.f); 
-    toa.fill(0.f);
+    chargeColl.fill(0.f);
+    toa1.fill(0.f);
+    toa2.fill(0.f);
     for(size_t i=0; i<it->second.hit_info[0].size(); i++) {
-      //time of arrival
-      float finalToA = (it->second).hit_info[1][i];
-      while(finalToA < 0.f)  finalToA+=25.f;
-      while(finalToA > 25.f) finalToA-=25.f;
-      toa[i]=finalToA;
+
+      // --- Fluctuate the total number of photo-electrons
+      float Npe = CLHEP::RandPoissonQ::shoot(hre, (it->second).hit_info[0][i]);
+      if ( Npe < EnergyThreshold_ ) continue;
+
+
+      // --- Get the time of arrival and add a channel time offset
+      float finalToA1 = (it->second).hit_info[1][i] + ChannelTimeOffset_;
+      
+      if (  smearChannelTimeOffset_ > 0. ){
+	float timeSmearing = CLHEP::RandGaussQ::shoot(hre, 0., smearChannelTimeOffset_);
+	finalToA1 += timeSmearing;
+      }
+
+
+      // --- Calculate and add the time walk: the time of arrival is read in correspondence
+      //                                      with two thresholds on the signal pulse
+      std::array<float, 3> times = btlPulseShape_.timeAtThr(Npe/ReferencePulseNpe_,
+							    TimeThreshold1_*Npe_to_V_, 
+							    TimeThreshold2_*Npe_to_V_);
+
+
+      // --- If the pulse amplitude is smaller than TimeThreshold2, the trigger does not fire
+      if (times[1] == 0.) continue;
+
+      float finalToA2 = finalToA1 + times[1];
+      finalToA1 += times[0];
+
+
+      // --- Uncertainty due to the fluctuations of the n-th photon arrival time:
+      float sigma2_tot_thr1 = ScintillatorDecayTime2_*sigma2_pe(TimeThreshold1_,Npe);
+      float sigma2_tot_thr2 = ScintillatorDecayTime2_*sigma2_pe(TimeThreshold2_,Npe);
+
+
+      // --- Add in quadrature the uncertainty due to the SiPM timing resolution:
+      sigma2_tot_thr1 += SPTR2_/TimeThreshold1_;
+      sigma2_tot_thr2 += SPTR2_/TimeThreshold2_;
+
+
+      // --- Add in quadrature the uncertainties due to the SiPM DCR, 
+      //     the electronic noise and the clock distribution:
+      float slew2 = ScintillatorDecayTime2_/Npe/Npe;
+
+      sigma2_tot_thr1 += ( (DCRxRiseTime_ + SigmaElectronicNoise2_)*slew2 + SigmaClock2_ );
+      sigma2_tot_thr2 += ( (DCRxRiseTime_ + SigmaElectronicNoise2_)*slew2 + SigmaClock2_ );
+
+
+      // --- Smear the arrival times using the total uncertainty:
+      finalToA1 += CLHEP::RandGaussQ::shoot(hre, 0., sqrt(sigma2_tot_thr1));
+      finalToA2 += CLHEP::RandGaussQ::shoot(hre, 0., sqrt(sigma2_tot_thr2));
+
+
+      while(finalToA1 < 0.f)  finalToA1 += 25.f;
```

Comment:
@casarsa @lgray @bendavid  I think we need to understand the meaning of this scaling (also in the next lines), which was originally added by Lindsey as far as I know. I miss it at present, is the electronics really supposed to work in this way?